### PR TITLE
First PR - Failover Connector skeleton

### LINKED ISSUE #20766: New component: Failover Connector
### The purpose and use-cases of the new component

A connector that routes data based on the current health status of a downstream component, typically an exporter.

I have heard several users ask for the ability to send data to a backup exporter, if a primary exporter fails. I believe this could be implemented as a routing connector.

The user would specify at least one pipeline to which data would typically be routed. Additionally, the user must specify at least one backup pipeline or pipelines which would be used when an error is encountered.

Initially, I think the trigger for routing to a backup pipeline could be based on backpropogated errors, though this is not yet very robust (See https://github.com/open-telemetry/opentelemetry-collector/issues/7460). At a later time, I imagine this could be based on the health status of an exporter (See https://github.com/open-telemetry/opentelemetry-collector/issues/6344).

### Example configuration for the component

```yaml
receivers:
  foo:

exporters:
  bar/main:
  bar/backup:

connectors:
  failover:
    primary: logs/main
    secondary: logs/backup

service:
  pipelines:
    logs/in:
       receivers: [foo]
       exporters: [failover]
    logs/main:
      receivers: [failover]
      exporters: [bar/main]
    logs/backup:
      receivers: [failover]
      exporters: [bar/backup]
```

### Telemetry data types supported

traces->traces
metrics->metrics
logs->logs

### Is this a vendor-specific component?

- [ ] This is a vendor-specific component
- [ ] If this is a vendor-specific component, I am proposing to contribute this as a representative of the vendor.

### Sponsor (optional)

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by akats7**
This is the Part 1 PR for the Failover Connector (split according to the CONTRIBUTING.md doc)

Link to tracking Issue: #20766 

Testing: Added factory test

Note: Full functionality PR exists [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/27641) and will likely be refactored to serve as the part 2 PR

cc: @djaglowski @sethallen @MovieStoreGuy

### CODE REVIEW

**Review by djaglowski — COMMENTED** [review]
Thanks @akats7.

I've done a cursory review of the boilerplate files. I'll circle back to the actual code included in the PR shortly.

I noted a couple nits, and we need some additional boilerplate changes as well:
- Please run `make chlog-new` and fill in the generate changelog file. 
- Please run `make gendependabot` to update `dependabot.yaml`
- Please add the component to `.github/CODEOWNERS`. Typically you and I would be owners for the component but you'll have to become a member of the OTel org to do so. For now, please just add me and plan on applying once we've merged a few PRs for this component.

File: connector/failoverconnector/README.md
Comment (djaglowski) [inline connector/failoverconnector/README.md:45]: Since these are at the same priority level, I wonder if it would be a bit clearer to new users if we renamed "fourth" to "also_first" or anything to that effect. It seems like the current name implies that it's somehow treated as a different priority.

File: connector/failoverconnector/README.md
Comment (djaglowski) [inline connector/failoverconnector/README.md:48]: This seems sufficiently descriptive to me. WDYT?
```suggestion
    retry_interval: 5m
```

File: connector/failoverconnector/README.md
Comment (djaglowski) [inline connector/failoverconnector/README.md:44]: We've got a mix of terms throughout this doc. "priority", "priority level" and "level". IMO "priority level" is the most clear. Maybe?
```suggestion
    priority_levels:
```

File: connector/failoverconnector/README.md
Comment (djaglowski) [inline connector/failoverconnector/README.md:71]: Where possible, it's helpful to use the exact config keys so the user doesn't have to assume what is intended.
```suggestion
The connector intakes a list of `priority_levels` each of which can contain multiple pipelines. If any pipeline at a stable level fails, the level is considered unhealthy and the connector will move down one priority level and route all data to the new level (assuming it is stable).
```

File: connector/failoverconnector/README.md
Comment (djaglowski) [inline connector/failoverconnector/README.md:73]:
```suggestion
The connector will periodically try to reestablish a stable connection with the higher priority levels. `retry_interval` will be the frequency at which the connector will try to iterate through all unhealthy higher priority levels while `retry_gap` is how long it will wait after a failed retry at one level before retrying the next level (if retry gap is 2m, after trying to reestablish level 1, it will wait 2m before trying level 2) It will retry a maximum of one unhealthy level before returning to the current stable level.)
```

File: connector/failoverconnector/README.md
Comment (djaglowski) [inline connector/failoverconnector/README.md:50]: This is documented below as "max_retries", which I think is clearer.
```suggestion
    max_retries: 10
```

File: connector/failoverconnector/README.md
Comment (djaglowski) [inline connector/failoverconnector/README.md:49]: Is "interval" accurate for how you expect this timing to behave? We use this term in a lot of similar situations, so I'd suggest it here as well unless there's a nuance to the behavior which makes it inaccurate.
```suggestion
    retry_interval: 1m
```

File: connector/failoverconnector/README.md
Comment (djaglowski) [inline connector/failoverconnector/README.md:74]:
```suggestion
There is a `max_retries` config param as well that will track how many retries have occurred at each level, and once the max is hit, it will no longer retry that priority level.
```

File: connector/failoverconnector/logs.go
Comment (djaglowski) [inline connector/failoverconnector/logs.go:42]: We can save this for a subsequent PR but I think we should validate the contents of all priority levels (i.e. instantiate each consumer) within the constructor, rather than defer until later. This will allow us to fail quickly on invalid input.

Comment (akats7) [reply connector/failoverconnector/logs.go:42]: Yep that's the way it works in the larger PR, I just removed it to avoid adding a mocked method on the failoverrouter

`	if !ok {
		return nil, errors.New("consumer is not of type LogsRouter")
	}

	failover := newFailoverRouter(lr.Consumer, config)
	err := failover.registerConsumers()
	if err != nil {
		return nil, err
	}`

File: connector/failoverconnector/README.md
Comment (akats7) [reply connector/failoverconnector/README.md:44]: Agreed

File: connector/failoverconnector/README.md
Comment (akats7) [reply connector/failoverconnector/README.md:48]: Agreed

File: connector/failoverconnector/logs.go
Comment (djaglowski) [reply connector/failoverconnector/logs.go:42]: Ok, sounds good. This is fine for this PR then.

File: connector/failoverconnector/README.md
Comment (akats7) [reply connector/failoverconnector/README.md:49]: So the interval is the time between the cycles of starting over from the highest priority pipeline down to the current stable one while gap is how long it waits between two levels in a single interval’s timeframe. If there’s a naming convention that aligns better to other similar uses, I’m all for it.

**Comment by crobert-1** [comment]
Just a housekeeping item: Can remove the word `Resolves` from this PR description line? 
```Link to tracking Issue: Resolves #20766```

Including the word `Resolves` here will close the issue, and I don't think we want to do that until the initial implementation is complete and merged.

**Comment by akats7** [comment]
> Just a housekeeping item: Can remove the word `Resolves` from this PR description line? `Link to tracking Issue: Resolves #20766`
> 
> Including the word `Resolves` here will close the issue, and I don't think we want to do that until the initial implementation is complete and merged.

Ah right, good call

File: connector/failoverconnector/README.md
Comment (djaglowski) [reply connector/failoverconnector/README.md:49]: Hmm, what's the reason for needing to add a delay here? Why not just immediately sent to the next priority after we drop down a level?

File: connector/failoverconnector/README.md
Comment (djaglowski) [inline connector/failoverconnector/README.md:37]: Typically we try to provide reasonable defaults wherever possible, in order to make it as simple as possible to get started with the component. These fields all seem like they could have reasonable defaults.

Comment (akats7) [reply connector/failoverconnector/README.md:37]: The idea is that it should retry a maximum of one unhealthy level before returning to the stable level to avoid back pressure through the collector. Especially with retries enabled on the exporters it could take several seconds for each level that is retried. If we’re on a low enough level and each higher priority level needs to be retried prior to returning to stability, that timeframe could become significant very quickly.

File: connector/failoverconnector/README.md
Comment (akats7) [reply connector/failoverconnector/README.md:49]: The idea is that it should retry a maximum of one unhealthy level before returning to the stable level to avoid back pressure through the collector. Especially with retries enabled on the exporters it could take several seconds for each level that is retried. If we’re on a low enough level and each higher priority level needs to be retried prior to returning to stability, that timeframe could become significant very quickly.

File: connector/failoverconnector/README.md
Comment (akats7) [reply connector/failoverconnector/README.md]: Addressed

File: connector/failoverconnector/README.md
Comment (akats7) [reply connector/failoverconnector/README.md]: Addressed

File: connector/failoverconnector/README.md
Comment (akats7) [reply connector/failoverconnector/README.md]: Addressed

File: connector/failoverconnector/README.md
Comment (akats7) [reply connector/failoverconnector/README.md]: Addressed

File: .github/CODEOWNERS
Comment (crobert-1) [inline .github/CODEOWNERS:37]:
```suggestion
connector/failoverconnector/                                            @open-telemetry/collector-contrib-approvers @djaglowski @fatsheep9146
```

It looks @fatsheep9146 [volunteered](https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/20766#issuecomment-1501036678) in the original issue to be the sponsor for this, so we can as code owner.

File: connector/failoverconnector/README.md
Comment (akats7) [reply connector/failoverconnector/README.md:49]: @djaglowski, any thoughts?

Comment (djaglowski) [reply connector/failoverconnector/README.md:49]: Ok, if I'm understanding correctly, any level which fails a retry will immediately end all retries until the next interval. However, if a retry is successful, we will both 1) reset the current level to the successful level, and 2) continue to retry any lower levels - and this gap is the time we wait until we retry the next level?

Comment (akats7) [reply connector/failoverconnector/README.md:49]: @djaglowski  Let me give an example:

connectors:
  failover:
    priority:
      - [traces/first]
      - [traces/second]
      - [traces/third]
      - [traces/fourth]
    retry_interval: 10m
    retry_gap: 1m
    max_retry: 10

Lets say we are currently on level 4 on the `priority_level` list. At the start of the retry_interval, the connector will try to reestablish the pipeline on level 1 (trace/first). If it fails, the connector will return to level 4 (traces/fourth) and wait the 1m as the retry_gap, when that 1m passes it will now retry level 2 (traces/second) and if that fails will first return to level 4 before waiting another 1m until trying level 3. Once it tries level 3 and it fails, it will return to level 4 and wait the 10m retry_interval again before repeating the process.

The idea for the retry_gap and returning to the _stable level_ between the _retry level_ is to make sure that the connector has a chance to export data without introducing a significant time gap or significant backpressure.

The idea for setting a retry_interval is to avoid constant retries, and allow for whatever needs to be resolved downstream to be resolved (e.g., a collector gateway that is down)

To answer you other question, if a retry is successful then yes the retried level becomes the stable level, and the connector will continue to retry any lower levels that haven't exceeded the `max_retries`

Comment (djaglowski) [reply connector/failoverconnector/README.md:49]: Thanks for clarifying @akats7. It seems worth capturing more detail in the readme description. Perhaps this example would be useful as a standalone section and then you can refer to it from the short descriptions in the table.

Comment (akats7) [reply connector/failoverconnector/README.md:49]: Sure thing, will add that to the readme.

File: connector/failoverconnector/README.md
Comment (fatsheep9146) [inline connector/failoverconnector/README.md:94]: > If a retry is successful then the retried level becomes the stable level, and the connector will continue to retry any lower levels that haven't exceeded the `max_retries`.

I did not get the meaning of the `lower levels`. You saild `max_retries` denotes "the maximum retries per level". Why is this changed to lower levels here?

File: connector/failoverconnector/config.go
Comment (fatsheep9146) [inline connector/failoverconnector/config.go:4]: I think this should be 
// import "github.com/open-telemetry/opentelemetry-collector-contrib/connector/failoverconnector"

File: connector/failoverconnector/config.go
Comment (fatsheep9146) [inline connector/failoverconnector/config.go:39]: This function should be tested

File: connector/failoverconnector/factory.go
Comment (fatsheep9146) [inline connector/failoverconnector/factory.go:24]: should use `metadata.Type`

File: connector/failoverconnector/factory.go
Comment (fatsheep9146) [inline connector/failoverconnector/factory.go:26]: should use `metadata.MetricsToMetricsStability `

**Comment by fatsheep9146** [comment]
failed checks should be fixed @akats7

**Comment by akats7** [comment]
> failed checks should be fixed @akats7

Hi @fatsheep9146, I believe they should all be resolved, looks like theres something going on with govulncheck in the rest of the pipeline

**Comment by akats7** [comment]
Hey @djaglowski, I believe everything has been addressed.

**Comment by fatsheep9146** [comment]
yes, the failed test seems related to #29109

**Comment by fatsheep9146** [comment]
I suggest we wait for #29109 merged, and the rebase this pr again, since some checks are skipped due to this problem.

**Comment by akats7** [comment]
Hi @djaglowski @fatsheep9146, all checks are passing

**Review by djaglowski — APPROVED** [review]

**Comment by djaglowski** [comment]
I've tagged this as ready to merge but will hold off while contrib release is in progress.

**Review by fatsheep9146 — APPROVED** [review]

**Pull request merged**
